EIPS/eip-2935.md

```diff
@@ -26,22 +27,56 @@ Additional secondary motivations include:
 
 | Parameter | Value |
 | - | - |
-| `FORK_BLKNUM` | TBD |
+| `FORK_TIMESTAMP` | TBD |
 | `HISTORY_STORAGE_ADDRESS` | `0xfffffffffffffffffffffffffffffffffffffffe`|
+| `HISTORY_SERVE_WINDOW` | `256` |
 
-At the start of processing any block where `block.number > FORK_BLKNUM` (ie. before processing any transactions), run `sstore(HISTORY_STORAGE_ADDRESS, block.number - 1, block.prevhash)`.
+At the start of processing any block where `block.timestamp > FORK_TIMESTAMP` (ie. before processing any transactions), update the history in the following way:
 
-When `block.number > FORK_BLKNUM + 256`, change the logic of the `BLOCKHASH` opcode as follows: if `FORK_BLKNUM <= arg < block.number`, return `sload(HISTORY_STORAGE_ADDRESS, arg)`. Otherwise return 0.
+```python
+def process_block_hash_history(block :Block, state: State):
+    if block.timestamp >= FORK_TIMESTAMP:
+        state.insert_slot(HISTORY_STORAGE_ADDRESS, block.number-1, block.parent.hash)
+
+    # If this is the first fork block, add the parent's direct 255 ancestors as well
+    if block.parent.timestamp < FORK_TIMESTAMP:
+        ancestor = block.parent
+        for i in range(HISTORY_SERVE_WINDOW - 1):
+            # stop at genesis block
+            if ancestor.number == 0:
+                break
+
+            ancestor = ancestor.parent
+            state.insert_slot(HISTORY_STORAGE_ADDRESS, ancestor.number, ancestor.hash)
+```
+
+Note that if this is the fork block, then it persists the additional requisite history that could be needed while resolving `BLOCKHASH` opcode for all of the `HISTORY_SERVE_WINDOW` ancestors (up untill genesis).
+
+For resolving the `BLOCKHASH` opcode this fork onwards (`block.timestamp > FORK_TIMESTAMP`), switch the logic to:
+
+```python
+def resolve_blockhash(block: Block, state: State, arg: uint64):
+  assert(arg < block.number)
```

Comment:
This does not match current `BLOCKHASH`. If `arg >= block.number` then return 0 (push 0 to stack)

Our code:

```typescript
      const number = runState.stack.pop()

      const diff = runState.interpreter.getBlockNumber() - number
      // block lookups must be within the past 256 blocks
      if (diff > BIGINT_256 || diff <= BIGINT_0) {
        runState.stack.push(BIGINT_0)
        return
      }
```

Reply:
yeah but that's precisely why we want to use the contract: be able to get further history than 256 blocks.

Reply:
Yes, the point is about when you request the current block number hash, or a block number higher, I would assume that `assert` throws? (And thus push nothing to stack and exit call frame?)

Reply:
applied the feedback